core/primitives/src/errors.rs

```diff
@@ -1,8 +1,53 @@
+use crate::serialize::u128_dec_format;
 use crate::types::{AccountId, Balance, Nonce};
 use borsh::{BorshDeserialize, BorshSerialize};
 use near_crypto::PublicKey;
+use serde::{Deserialize, Serialize};
 use std::fmt::Display;
 
+use near_rpc_error_macro::RpcError;
+use near_vm_errors::VMError;
+
+/// Error returned in the ExecutionOutcome in case of failure.
+#[derive(
+    BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize, RpcError,
+)]
+#[rpc_error_variant = "TxExecutionError"]
+pub enum ExecutionError {
+    Action(ActionError),
+    InvalidTx(InvalidTxError),
+}
+
+impl Display for ExecutionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        match self {
+            ExecutionError::Action(e) => write!(f, "{}", e),
+            ExecutionError::InvalidTx(e) => write!(f, "{}", e),
+        }
+    }
+}
+
+impl From<ActionError> for ExecutionError {
+    fn from(error: ActionError) -> Self {
+        ExecutionError::Action(error)
+    }
+}
+
+impl From<InvalidTxError> for ExecutionError {
+    fn from(error: InvalidTxError) -> Self {
+        ExecutionError::InvalidTx(error)
+    }
+}
+
+/// Error returned from `Runtime::apply`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum RuntimeError {
+    UnexpectedIntegerOverflow,
+    InvalidTxError(InvalidTxError),
+    StorageError(StorageError),
+    BalanceMismatch(BalanceMismatchError),
```

Comment:
There is some inconsistency in naming. In other places, you use `InvalidTx` enum variant (without `Error` suffix). Naming is hard :smile: (I am still not sure which one I like better `InvalidTx` or `InvalidTxError`)

Reply:
renamed to InvalidTxError